nwb_conversion_tools/datainterfaces/axonadatainterface.py

```diff
@@ -0,0 +1,136 @@
+                header[key] = ' '.join(parts[1:])
+
+    return header
+
+
+def read_iso_datetime(set_file):
+    """
+    Creates datetime object (y, m, d, h, m, s) from .set file header
+    """
+    with open(set_file, 'r', encoding='cp1252') as f:
+        for line in f:
+            if line.startswith('trial_date'):
+                date_string = re.findall(r'\d+\s\w+\s\d{4}$', line)[0]
+            if line.startswith('trial_time'):
+                time_string = line[len('trial_time')+1::].replace('\n', '')
+
+    return datetime.datetime.strptime(date_string + ', ' + time_string,
+                                      "%d %b %Y, %H:%M:%S").isoformat()
+
+
+class AxonaRecordingExtractorInterface(BaseRecordingExtractorInterface):
+    """
+    Primary data interface class for converting a AxonaRecordingExtractor
+    """
+
+    RX = se.AxonaRecordingExtractor
+
+    @classmethod
+    def get_source_schema(cls):
+        source_schema = {
+            'required': ['filename'],
+            'properties': {
+                'filename': {
+                    'type': 'string',
+                    'format': 'file',
+                    'description': 'Path to Axona files.'
+                }
+            },
+            'type': 'object',
+            'additionalProperties': True
+        }
+        return source_schema
+
+    def __init__(self, **source_data):
+        super().__init__(**source_data)
+
+    def get_metadata(self):
+        """
+        Auto-fill metadata where possible. Must comply with metadata schema.
+        """
+
+        # Extract information for specific parameters from .set file
+        # TODO not all are currently used or part of the metadata_schema
+        params_of_interest = [
+            'experimenter',
+            'comments',
+            'duration',
+            'sw_version',
+            'tracker_version',
+            'stim_version',
+            'audio_version'
+        ]
+        set_file = self.source_data['filename']+'.set'
+        par = parse_generic_header(set_file, params_of_interest)
+
+        # Add available metadata
+        # NOTE that this interface is meant to be used within an NWBconverter,
+        # which contains a much larger metadata_schema. As such a datainterface
+        # metadata by itself does not necessarily validate with its own schema!
+        metadata = super().get_metadata()
+        metadata['NWBFile'] = dict(
+            session_start_time=read_iso_datetime(set_file),
+            session_description=par['comments'],
+            identifier=''.join(random.choices(string.ascii_uppercase +
+                                              string.digits, k=12)),
+            experimenter=[par['experimenter']]
+        )
+
+        metadata['Ecephys'] = dict(
+            Device=dict(
+                    name="Axona",
+                    description="Axona DacqUSB, sw_version={}".format(
+                        par['sw_version']),
+                    manufacturer="Axona"
```

Comment:
Ah hah, interesting. This won't actually get added to the NWBFile since we don't have it built in to do so. Could you raise an issue on spikeextractors and @ me to get a PR that allows it to get passed to the metadata?

Reply:
@CodyCBakerPhD 

I currently have the following code:
```
metadata['NWBFile'] = dict(
            session_start_time=read_iso_datetime(set_file),
            session_description=par['comments'],
            session_duration=par['duration']+'s',
            experimenter=[par['experimenter']]
)
```

And they do seem to get added as "defaults", with the exception of `session_description` - is that what you meant?:

```
{
  "NWBFile": {
    "required": [
      "session_description",
      "identifier",
      "session_start_time"
    ],
    "properties": {
      "session_description": {
        "type": "string",
        "format": "long",
        "description": "a description of the session where this data was generated",
        "default": ""
      },
      "identifier": {
        "type": "string",
        "description": "a unique text identifier for the file",
        "default": "8d4cf566-4041-4518-ada7-61a36b1466b9"
      },
      "session_start_time": {
        "type": "string",
        "description": "the start date and time of the recording session",
        "format": "date-time",
        "default": "2020-10-04T11:07:07"
      },
      "experimenter": {
        "type": "array",
        "items": {
          "type": "string",
          "title": "experimenter"
        },
        "description": "name of person who performed experiment",
        "default": [
          "Abid"
        ]
      },
     ...
```

Reply:
@CodyCBakerPhD this one took me a while to get my head around, but I think I understand what you are saying now. 

When I omit some of this information and follow for instance [neuroscope](https://github.com/catalystneuro/nwb-conversion-tools/blob/master/nwb_conversion_tools/datainterfaces/neuroscopedatainterface.py), then I am able to convert to an NWB file.

But if I leave the additional attributes, it crashes with the following error.

```
data_interface.run_conversion(nwbfile, metadata)
```

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-107-71ef00e12343> in <module>
----> 1 data_interface.run_conversion(nwbfile, metadata)

/mnt/d/spikeinterface/nwb-conversion-tools/nwb_conversion_tools/baserecordingextractorinterface.py in run_conversion(self, nwbfile, metadata, stub_test, use_times, save_path, overwrite, buffer_mb)
    124             recording = self.recording_extractor
    125 
--> 126         se.NwbRecordingExtractor.write_recording(
    127             recording=recording,
    128             nwbfile=nwbfile,

/mnt/d/spikeinterface/spikeextractors/spikeextractors/extractors/nwbextractors/nwbextractors.py in write_recording(recording, save_path, overwrite, nwbfile, buffer_mb, use_times, metadata, write_as_lfp, write_scaled)
   1045                 io.write(nwbfile)
   1046         else:
-> 1047             se.NwbRecordingExtractor.add_all_to_nwbfile(
   1048                 recording=recording,
   1049                 nwbfile=nwbfile,

/mnt/d/spikeinterface/spikeextractors/spikeextractors/extractors/nwbextractors/nwbextractors.py in add_all_to_nwbfile(recording, nwbfile, buffer_mb, use_times, metadata, write_as_lfp, write_scaled)
    910         )
    911 
--> 912         se.NwbRecordingExtractor.add_electrode_groups(
    913             recording=recording,
    914             nwbfile=nwbfile,

/mnt/d/spikeinterface/spikeextractors/spikeextractors/extractors/nwbextractors/nwbextractors.py in add_electrode_groups(recording, nwbfile, metadata)
    440                         )
    441                     )
--> 442                     se.NwbRecordingExtractor.add_devices(recording, nwbfile, metadata=new_device)
    443                     warnings.warn(f"Device \'{device_name}\' not detected in "
    444                                   "attempted link to electrode group! Automatically generating.")

/mnt/d/spikeinterface/spikeextractors/spikeextractors/extractors/nwbextractors/nwbextractors.py in add_devices(recording, nwbfile, metadata)
    376                 )
    377             )
--> 378         assert all([isinstance(x, dict) for x in metadata['Ecephys']['Device']]), \
    379             "Expected metadata['Ecephys']['Device'] to be a list of dictionaries!"
    380 

AssertionError: Expected metadata['Ecephys']['Device'] to be a list of dictionaries!
```

When going through this I found that in spikeextractors, `nwbextractors.add_electrode_groups` 
it looks for the dictionary key "device_name" in `metadata['Ecephys']['ElectrodeGroup']`, which
contains "device" instead (for each list item). Since the key is not found, it assigns the default
value of 'Device' and after a few more steps it tries to add a "default" dictionary for device,
but this is not compatible with the schema. 

```
defaults = dict(
    name="Electrode Group",
    description="no description",
    location="unknown",
    device_name="Device"
)

...

for grp in metadata['Ecephys']['ElectrodeGroup']:
    device_name = grp.get('device_name', defaults['device_name'])
    if grp.get('name', defaults['name']) not in nwbfile.electrode_groups:
        if device_name not in nwbfile.devices:
            new_device = dict(
                Ecephys=dict(
                    Device=dict(
                        name=device_name
                    )
                )
            )
            se.NwbRecordingExtractor.add_devices(recording, nwbfile, metadata=new_device)
```

It might be sufficient to change `device_name = grp.get('device_name', ...)` to
`device_name = grp.get('device', ...)`.